### ISSUE TITLE
Default theme rename - and issues

### ISSUE THREAD

**Issue by philwareham** [textpattern/textpattern#1101]
I will rename the `default` theme to `four-point-seven` shortly as per Pete's good suggestion. But while playing with some name combinations  noted the following things that might need to be fixed...

1. There is a file `/textpattern/setup/data/txp_skin.xml` that seems to directly reference the theme. Shouldn't that be dropped in favour of the `manifest.json` file within the theme itself? I notice there is not any entry for the 'empty' theme we also ship.
2. If the theme's folder has a point `.` in the folder name, the setup elect list fails to see it. Not sure if this is a quirk only in the setup or if this issue affects the standard themes directory too? Anyway, it'd need to be fixed.
3. How do we tell setup which theme is selected by default in the `<select>` list. I think it just does it alphabetically at the moment but since renaming `default` to `txp_47` (temporality) the first item now selected in the list is the empty theme - not desirable.

Cheers!

**Comment by Bloke** [comment]
Good calls, will fix all these and check out the decimal point issue in the name. Probably, the method of stripping the extension off is wonky.

**Comment by philwareham** [comment]
Well I have renamed the theme now. The issues above still need consideration though. Cheers!

**Comment by petecooper** [comment]
Addendum: automated install on demo site is pulling down a theme called `default` as before, with empty forms, pages, and no content.

Pros: I found another missing gTxt. Hurrah!
Cons: it's a bit…sparse front-side.

**Comment by Bloke** [comment]
Best guesses:

1. It's picking the 'first' theme in the list, which is now `empty` (alphabetically) but calling it `default` when it's installed.
2. It's being told to use `default` but that doesn't exist any more so it's choosing the first in the list. See above.
3. Did you update the installer to use `four-point-seven` as the theme name to install?

Failing any of those, when I fix the default selector for manual install, I'll look at this too.

**Comment by petecooper** [comment]
Hey @Bloke – it's 1., and possibly 2. I haven't explicitly mentioned which front-side theme to use in the JSON (I'll fix that now).

**JSON**

```
$ cat /var/www/vhosts/textpattern.co/dev-demo/staging/setup.json
{
    "site": {
        "lang": "en",
        "siteurl": "",
        "theme": "",
        "public_theme": "",
        "datadir": ""
    },
    "mysql": {
        "user": "devdemo",
        "pass": "<snipped>",
        "host": "127.0.0.1",
        "db": "devdemo",
        "table_prefix": ""
    },
    "user": {
        "realname": "Textpattern Demo Robot",
        "email": "domo-arigato-mr-roboto@example.com",
        "name": "textpattern-robot",
        "pass": "<snipped>"
    }
}
```

**Output**

```
$ sudo /var/www/vhosts/textpattern.co/dev-demo/staging/dev-demo.sh
mysql: [Warning] Using a password on the command line interface can be insecure.
Cloning into '/var/www/vhosts/textpattern.co/dev-demo/staging/github'...
remote: Counting objects: 476, done.
remote: Compressing objects: 100% (417/417), done.
remote: Total 476 (delta 100), reused 145 (delta 21), pack-reused 0
Receiving objects: 100% (476/476), 1.35 MiB | 0 bytes/s, done.
Resolving deltas: 100% (100/100), done.
Checking connectivity... done.
[OK]	Connected
[OK]	Using devdemo. (utf8mb4)
[OK]	Creating database tables
[OK]	Lang: 'en'
[OK]	Front-side theme: '/setup/themes/empty'
[OK]	CSS: 'default'
[OK]	Form: 'article/default'
[OK]	Form: 'comment/comment_form'
[OK]	Form: 'comment/comments'
[OK]	Form: 'comment/comments_display'
[OK]	Form: 'file/files'
[OK]	Form: 'link/plainlinks'
[OK]	Page: 'default'
[OK]	Page: 'error_default'
[OK]	That went well!
mysql: [Warning] Using a password on the command line interface can be insecure.
```

So, yes – `empty` comes before `four-point-seven` (I knew this would bloody bite me…). Well deduced! I'll open a fresh issue for the missing-but-not-really 2x gTxt.

**Comment by petecooper** [comment]
Actually, I spoke too soon – even with explicitly specifying `four-point-seven`…

```
    "site": {
        "lang": "en",
        "siteurl": "",
        "theme": "",
        "public_theme": "four-point-seven",
        "datadir": ""
    },
    "mysql": {
        "user": "devdemo",
        "pass": "<snipped>",
        "host": "127.0.0.1",
        "db": "devdemo",
        "table_prefix": ""
    },
    "user": {
        "realname": "Textpattern Demo Robot",
        "email": "domo-arigato-mr-roboto@example.com",
        "name": "textpattern-robot",
        "pass": "<snipped>"
    }
}
```

…it still grabs `empty` instead:

```
$ sudo /var/www/vhosts/textpattern.co/dev-demo/staging/dev-demo.sh
mysql: [Warning] Using a password on the command line interface can be insecure.
Cloning into '/var/www/vhosts/textpattern.co/dev-demo/staging/github'...
remote: Counting objects: 476, done.
remote: Compressing objects: 100% (417/417), done.
remote: Total 476 (delta 100), reused 145 (delta 21), pack-reused 0
Receiving objects: 100% (476/476), 1.35 MiB | 0 bytes/s, done.
Resolving deltas: 100% (100/100), done.
Checking connectivity... done.
[OK]	Connected
[OK]	Using devdemo. (utf8mb4)
[OK]	Creating database tables
[OK]	Lang: 'en'
[OK]	Front-side theme: '/setup/themes/empty'
[OK]	CSS: 'default'
[OK]	Form: 'article/default'
[OK]	Form: 'comment/comment_form'
[OK]	Form: 'comment/comments'
[OK]	Form: 'comment/comments_display'
[OK]	Form: 'file/files'
[OK]	Form: 'link/plainlinks'
[OK]	Page: 'default'
[OK]	Page: 'error_default'
[OK]	That went well!
mysql: [Warning] Using a password on the command line interface can be insecure.
```

**Comment by Bloke** [comment]
Cool, thanks for testing. I'll add it to the list!

**Comment by petecooper** [comment]
@Bloke - top man, thank you very much. You want me to bounce a new issue for it or keep it here?

**Comment by Bloke** [comment]
Nah leave it here and I'll pick it up when I do the rest.

**Comment by Bloke** [comment]
Hopefully, 83b84e14 fixes both the manual and auto-theme selector, plus permits dots and hyphens to be used. We should really use the full regex that we use for validating theme names, but we can worry about that later.

Please test and report back.

P.S. still haven't looked into the `txp_skin.xml` issue. Will do that next.

**Comment by petecooper** [comment]
Thanks for your work on this @Bloke - really appreciated. I have some feedback on the current dev demo situation, which is using the current GitHub `dev` branch (every three hours) and a straightforward auto-installer. This *may* be better suited to a separate issue, I'll be guided by you on that. I may just be doing something wrong.

The dev site is still content-free, zero articles and forms are blank. The installed theme (singular) is named `default` in the admin-side, which is at odds with the `empty` theme the installer says it grabs.

Here's the (sanitised) JSON:

```
{
    "site": {
        "lang": "en",
        "siteurl": "",
        "theme": "",
        "public_theme": "four-point-seven",
        "datadir": ""
    },
    "mysql": {
        "user": "devdemo",
        "pass": "<SNIPPYSNIPPY>",
        "host": "127.0.0.1",
        "db": "devdemo",
        "table_prefix": ""
    },
    "user": {
        "realname": "Textpattern Demo Robot",
        "email": "domo-arigato-mr-roboto@example.com",
        "name": "textpattern-robot",
        "pass": "<SNIPPYSNIPPY>"
    }
}
```

And here's the output from the auto-installer:

```
$ sudo /var/www/vhosts/textpattern.co/dev-demo/staging/dev-demo.sh
mysql: [Warning] Using a password on the command line interface can be insecure.
Cloning into '/var/www/vhosts/textpattern.co/dev-demo/staging/github'...
remote: Counting objects: 476, done.
remote: Compressing objects: 100% (418/418), done.
remote: Total 476 (delta 99), reused 138 (delta 20), pack-reused 0
Receiving objects: 100% (476/476), 1.35 MiB | 0 bytes/s, done.
Resolving deltas: 100% (99/99), done.
Checking connectivity... done.
[OK]	Connected
[OK]	Using devdemo. (utf8mb4)
[OK]	Creating database tables
[OK]	Lang: 'en'
[OK]	Front-side theme: '/setup/themes/empty'
[OK]	CSS: 'default'
[OK]	Form: 'article/default'
[OK]	Form: 'comment/comment_form'
[OK]	Form: 'comment/comments'
[OK]	Form: 'comment/comments_display'
[OK]	Form: 'file/files'
[OK]	Form: 'link/plainlinks'
[OK]	Page: 'default'
[OK]	Page: 'error_default'
[OK]	That went well!
mysql: [Warning] Using a password on the command line interface can be insecure.
```

Note the `setup/themes/empty` theme being imported. Note the theme name on the admin-side:

![safariscreensnapz001](https://user-images.githubusercontent.com/226578/34874722-2628c7d0-f792-11e7-8699-b3238e926377.png)

Follow-up: leaving the `public_theme` empty in the auto-installer JSON has the same effect: grabs `empty` in the installer, calls it `default` in the admin-side.

**Comment by bloatware** [comment]
I don't think `empty` theme is a good idea at all - what is it for? Create a new theme if you want to start from scratch, that's all.

**Comment by petecooper** [comment]
@bloatware - my understanding is that you can elect to choose the `empty` theme during setup and it will install the theme scaffold, minus the content, and crucially minus the default content (article, comment, links, and so on).

**Comment by bloatware** [comment]
Isn't that what "Create new theme" button does? BTW, I get
```
Notice "Undefined index: website"
in D:\wamp\www\textpattern\textpattern\vendors\Textpattern\Skin\Main.php at line 223.
```
As for the default content, we could add an option to create it (or not) on setup.

**Comment by petecooper** [comment]
My 2c: personally, I don't have any strong feelings either way. I made a [purely scaffold Textpattern front-side theme](https://github.com/petecooper/textpattern-front-side-theme-scaffold) back in April 2017 and I've updated to match the current file tree. The simplicity of installing Textpattern without the default content is refreshing to me, and it's surprisingly fast how 10-plus years of muscle memory can disappear!

With @makss's auto-installer and the huge amount of evolutionary changes coming in 4.7.0, I feel spoiled – it's _so easy_ and incredibly fast to get rolling quickly now. From a user/web dev point of view, I don't really care _how_ we go about a boilerplate Textpattern instance, as long as it's somewhat straightforward, can be replicated and is (ideally) documented.

The `empty` theme works for me, personally. Right now it's a bit strange since it's the first in the list alphabetically now that `default` has been renamed `four-point-seven` (you can blame me for that!), which is causing some auto-installer oddness (see [above](https://github.com/textpattern/textpattern/issues/1101#issuecomment-357223922)).

…I did think renaming `empty` to `zero` might be a kludge, but I've caused enough trouble with theme names in the last 2 days!

**Comment by phiw13** [comment]
@bloatware 

> As for the default content, we could add an option to create it (or not) on setup.

Isn’t it what currently happens at the end of setup? You have the choice between the (default) “four-point-seven“ (with an article etc) and an “empty” theme. The only thing the latter does is create placeholders for the (currently required) `pages` and `forms` (+ a stylesheet). That works for me.

I might argue with the choice of name maybe — “empty” ? But I’m currently short of ideas as for an alternative.

**Comment by Bloke** [comment]
Thanks, @petecooper . I just found that annoying 'default' thing getting injected into the database during manual setup too. I've hacked around in my local code and something weird is going on when the setup routine is calling `theme->import()`. It seems it wants to be given assets at that time, rather than being given a directory name and reading them from there.

There's a possibility this is what's happening:

* Choose four-point-seven during setup, which actually passes `setup/themes/four-point-seven` to the final step _not just_ the theme name. Side note: the select box in the installer has the path in it, which makes it pretty wide and it nudges to the left, out of line with other boxes. @philwareham It might need a wider area devoted to it, or content truncated until clicked.
* Installer kicks off, removes the path (because the Theme engine needs just the name to instantiate it). This it receives `four-point-seven` and looks in the themes dir for it.
* Doesn't find it there. Shrugs. Uses empty content for assets.
* Also doesn't find a manifest so it uses default names.

That's my best guess right now. There's probably a method I need to call during instantiation to tell it to look in the `setup/themes` folder instead but I haven't got round to trawling the code to find out yet. If anyone fancies taking a crack at fixing this today, I won't be offended :-) There are a bunch of FIXME parts to the installer which I've altered in my local dev (e.g. `Textpattern/Skin/Skin` has been renamed `Textpattern/Skin/Main`) but nothing life threatening that I can't undo if someone finds a decent solution in the meantime. If not, I'll carry on hacking away tonight.

**Comment by petecooper** [comment]
Nice sleuthing as always, @Bloke - I'll take another look with fresh eyes after lunch (maybe soup, maybe not).

**Comment by Bloke** [comment]
And, yes, `empty` is simply that: no content, no template code, no nothing but the bare minimum scaffolding to satisfy Txp's need for some default Forms.

Textpattern requires _one theme at minimum_ to be defined, so this is a shortcut. Otherwise you would have to do this:

1. Select four-point-seven during installation.
2. Log in.
3. Create new empty theme.
4. Reassign all Sections to use it.
5. Delete four-point-seven theme from the DB.

OR:

1. Supply your own empty or populated theme during install with or without content, plugins, prefs, etc.

For people that want an absolute blank canvas on which to build, `empty` is a great option. As for a cool name, well, we can leave that to Pete ;-)

**Comment by petecooper** [comment]
I like `zero`, actually, having given it some thought. No content, no nothing. And it's at the end of the list until something more zz-y comes along.

**Comment by bloatware** [comment]
I'm not sure that displaying blank public site before you fill the `empty` theme is better than `four-point-seven` (SEO, perhaps?). And it's only one `default` section to reassign when you are ready.

**Comment by Bloke** [comment]
Yeah, there's not much in  it. I just see it as a way to really start from scratch or drop a new theme in from a third-party site. Like [HTML5up](http://html5up.net) for example.

Sure you could install `four-point-seven` then create a new theme from the UI and paste the template code in (or overwrite four-point-seven), tweak the result by adding txp tags and then delete existing content ready for your client stuff.

Or, if you know you're going to be doing this anyway (e.g. client says "I want you to build me a site that looks like  _Future Imperfect on HTML5up_") you could just set up a new Txp with the _zero_ theme and paste the template code straight in, knowing that when you view the site it's immediately wired up to show live.

Or you might already have a Txp-ised version of that template lying around so you could go one better and drop it in the themes directory prior to set up and load it in during installation.

Dunno. Each to their own. But I do see a complete blank slate as a useful shortcut for power users who want to stuff new templates in without having to delete anything first.

Either way, the setup import routine needs to be hooked up to the `Themes->import()` routine and do it properly. This has highlighted a bug in the setup process (and possibly the way themes are implemented that might require a minor tweak unless we can find the method to tell it to load from somewhere other than the `themes` directory) so this is a good stress test, regardless of whether we ultimately keep `zero` or not.

**Comment by jools-r** [comment]
Is it correct that on choosing an install with ‘Four Point Seven’ the Themes panel shows you four-point-seven but 0 Sections, 0 Pages, 0 Forms, 0 Styles, etc.? An export into a directory gives you the theme folder, manifest.json but not pages, forms or styles. I was wondering initially if this was a multisite anomaly, but I get this with the current dev branch, too.

**Comment by Bloke** [comment]
No, it's broken, sorry. Not pulling in the assets for reasons I still cannot fathom. It'll be something stupid I'm overlooking.

**Comment by petecooper** [comment]
I will gladly volunteer to be a [rubber duck](https://en.wikipedia.org/wiki/Rubber_duck_debugging) if that would help, @Bloke

**Comment by petecooper** [comment]
Also – @Bloke – I can run the automated setup with debug mode if you want a verbose install log. Any help?

**Comment by petecooper** [comment]
Thinking out loud for a sec…

The default theme .json file includes a `title`, but not a `name` field. Is the `name` value automatically inherited from the containing directory?

**Comment by Bloke** [comment]
Yeah, it's supposed to pick up the name from the directory, as your surmise. After a bit of hacking, it seems to be picking up the JSON stuff now, but not loading the theme assets for some reason from the setup directory. Annoying.

**Comment by petecooper** [comment]
Sure thing. I'm running this from a freshly-cloned 4.7.0-dev git repo:

```
php /var/www/vhosts/textpattern.co/dev-demo/live/textpattern/setup/setup.php --debug  --config="/var/www/vhosts/textpattern.co/dev-demo/staging/setup.json"
```

…the contents of the `setup.json` (I've sanitised passwords):

```json
$ cat /var/www/vhosts/textpattern.co/dev-demo/staging/setup.json
{
    "site": {
        "lang": "en",
        "siteurl": "",
        "theme": "",
        "public_theme": "Four Point Seven",
        "datadir": ""
    },
    "mysql": {
        "user": "devdemo",
        "pass": "<SNIPPED>",
        "host": "127.0.0.1",
        "db": "devdemo",
        "table_prefix": ""
    },
    "user": {
        "realname": "Textpattern Demo Robot",
        "email": "domo-arigato-mr-roboto@example.com",
        "name": "textpattern-robot",
        "pass": "<SNIPPED>"
    }
}
```

…and the zipped debug output is attached. If you'd prefer a gist, happy to oblige.

[debug.txt.zip](https://github.com/textpattern/textpattern/files/1629552/debug.txt.zip)

**Comment by Bloke** [comment]
Thanks, what happens if you specify the public theme name as its folder name `four-point-seven` in the `setup.json` file?

**Comment by petecooper** [comment]
If I use `four-point-seven`, so (excerpted)…

```json
"site": {
        "lang": "en",
        "siteurl": "",
        "theme": "",
        "public_theme": "four-point-seven",
        "datadir": ""
    },
```

…same thing as https://github.com/textpattern/textpattern/issues/1101#issuecomment-357525585 – grabs the `empty` theme (according to Terminal output), admin-side shows a single theme called `default`, version `4.7.0-dev`, no description.

I also tried the setup JSON with no theme specified, so (excerpted)…

```json
"site": {
        "lang": "en",
        "siteurl": "",
        "theme": "",
        "public_theme": "",
        "datadir": ""
    },
```

…same thing.

Two debug logs attached: `four-point-seven.txt.zip` is where the theme name was specified, and `notheme.txt.zip` was where no theme was specified.

[notheme.txt.zip](https://github.com/textpattern/textpattern/files/1629579/notheme.txt.zip)
[four-point-seven.txt.zip](https://github.com/textpattern/textpattern/files/1629580/four-point-seven.txt.zip)

**Comment by Bloke** [comment]
Thought as much, just wanted to make sure, thanks. I'll trawl the debug info and see if I can figure out where it's falling over.

**Comment by petecooper** [comment]
Thanks @Bloke - will be around this evening if you need second pair of eyes etc.

**Comment by Bloke** [comment]
Untested in unattended installs as yet, but I just [pushed a commit](https://github.com/textpattern/textpattern/commit/8bba469b51a4568e59c20b4e1da4c0f8ab274343) that allowed installation from locations other than the `themes` directory, and wired the setup process to use it.

If it works, we can probably remove the code noted in the todo in `setup_lib.php` since the skin system always exists from this version forward.

Hopefully this will also fix auto-install since it uses the same code. Test results welcome.

**Comment by philwareham** [comment]
Just done a clean install and seems to work (manual install). Four Point Seven is listed as the default selected theme during the installation process.

Not sure we really need the full location directory of the theme listed in that select list, it's prone to breaking the layout. I understand that it's there for if a user has put a theme with exactly the same name in the standard `themes` folder but for most users it's unhelpful cruft.

Would be better if the system checked whether the same theme name appears in both locations and then, if so, adds the directory details - if not, doesn't show the directory details.

**Comment by Bloke** [comment]
Yeah, I think removing that directory info unless necessary would be beneficial. I'll do that.

**Comment by petecooper** [comment]
Automated installer person checking in – hello!

@Bloke - thanks for your continued work with this, massively appreciated. I have some good news: the `default` theme with the inherited version number has now gone! Hurrah! I get a proper theme installed, now.

Unfortunately, whether I specify `"public_theme": "Four Point Seven",`, `"public_theme": "four-point-seven",` or `"public_theme": "",` it always seems to install `empty`, according to tty output and the admin-side.

So…again, thinking out loud…is `public_theme` correct? Perhaps it's ignored?

**Comment by philwareham** [comment]
I've changed the empty theme to `zero` now as discussed. Still like to see Pete's issues with the auto-install fixed without relying on alphabetical order.

**Comment by petecooper** [comment]
Happy to report that the demo site is back up and running with Four Point Zero theme since theme since https://github.com/textpattern/textpattern/commit/b411a153c9585d668e1f9209546741d08e5b41cf landed.

**Comment by Bloke** [comment]
Thanks for testing, @petecooper . Looks like there's some more work to do on the automated side of things so it picks up the value in the JSON file properly. Hmmm. Will try and fix that today.

As far as I'm aware, it should use the _name_ (directory name, not customer-friendly title) of the theme in the `public_theme` setting.

**Comment by petecooper** [comment]
If it helps, there doesn't appear to be any error reporting on the JSON, as far as I can tell. However, at the latest commit, using:

``` json
"public_theme": "non-existent theme tra la la",
```

…falls back to Four Point Seven, without error:

```
$ sudo /var/www/vhosts/textpattern.co/dev-demo/staging/dev-demo.sh
mysql: [Warning] Using a password on the command line interface can be insecure.
Cloning into '/var/www/vhosts/textpattern.co/dev-demo/staging/github'...
remote: Counting objects: 475, done.
remote: Compressing objects: 100% (417/417), done.
remote: Total 475 (delta 99), reused 142 (delta 20), pack-reused 0
Receiving objects: 100% (475/475), 1.35 MiB | 0 bytes/s, done.
Resolving deltas: 100% (99/99), done.
Checking connectivity... done.
[OK]	Connected
[OK]	Using devdemo. (utf8mb4)
[OK]	Creating database tables
[OK]	Lang: 'en'
[OK]	Prefs: 'data/core'
[OK]	Import: 'data/txp_category'
[OK]	Import: 'data/txp_link'
[OK]	Import: 'data/txp_section'
[OK]	Import: 'articles/articles.welcome'
[OK]	Front-side theme: '/setup/themes/four-point-seven'
[OK]	That went well!
mysql: [Warning] Using a password on the command line interface can be insecure.
```

**Comment by petecooper** [comment]
I can give you a `sudo` login shell if you want to try stuff, no worries.

**Comment by Bloke** [comment]
Penny dropped. It does work, it's just counter-intuitive right now. The problem? You need to put the _path_ to the public_theme in the config file.

The stupid thing is - and I think we're going to need to change this, but I can't think how at the moment - is that it's a relative path. But it has a leading slash and is relative to, ummm, well it depends:

* If you're installing from setup: `/setup/themes/theme-name-to-install`
* If you're installing from the themes dir: `/themes/theme-name-to-install`

That's pretty dim-witted in a way, but if you drop a custom theme in the `themes` directory and go through the manual setup process then use the DOM inspector to look at the options in the public theme selector in the last step, you'll see the `value` attribute does indeed have such paths. They're relative to, well, the 'type' of theme you're installing - from setup or from the public space. Madness.

For consistency, we should force the directory to be relative to docroot (and therefore it'll need to be `/textpattern/setup/themes/theme-name-to-install`). More to type. OR we get clever and just let you use the name, and try to figure out which one you mean, hoping that you don't have two identically-named themes in the `themes` dir and in `setup/themes`. Or, maybe if you do, we choose the public one over the setup one? Not sure.

Bottom line. It works if you put the sort-of-relative path in, but I'd like it to be more intelligent than it is right now. Will investigate options.

**Comment by petecooper** [comment]
*Damn*, you're good.

**Comment by petecooper** [comment]
OK, some feedback. Using the current 4.7.0-dev:

```
php /var/www/vhosts/textpattern.co/dev-demo/live/textpattern/setup/setup.php  --config="/var/www/vhosts/textpattern.co/dev-demo/staging/setup.json"
```
…with this in the JSON (valid theme):
```
"public_theme": "/setup/themes/four-point-seven",`
```
…it looks like `setup.php` grabs the correct theme:
```
$ sudo /var/www/vhosts/textpattern.co/dev-demo/staging/dev-demo.sh
mysql: [Warning] Using a password on the command line interface can be insecure.
Cloning into '/var/www/vhosts/textpattern.co/dev-demo/staging/github'...
remote: Counting objects: 475, done.
remote: Compressing objects: 100% (417/417), done.
remote: Total 475 (delta 99), reused 139 (delta 20), pack-reused 0
Receiving objects: 100% (475/475), 1.36 MiB | 1.48 MiB/s, done.
Resolving deltas: 100% (99/99), done.
Checking connectivity... done.
[OK]	Connected
[OK]	Using devdemo. (utf8mb4)
[OK]	Creating database tables
[OK]	Lang: 'en'
[OK]	Prefs: 'data/core'
[OK]	Import: 'data/txp_category'
[OK]	Import: 'data/txp_link'
[OK]	Import: 'data/txp_section'
[OK]	Import: 'articles/articles.welcome'
[OK]	Front-side theme: '/setup/themes/four-point-seven'
[OK]	That went well!
mysql: [Warning] Using a password on the command line interface can be insecure.
```
Next step – test `zero` theme by setting `"public_theme": "/setup/themes/zero",`…

```
$ sudo /var/www/vhosts/textpattern.co/dev-demo/staging/dev-demo.sh
mysql: [Warning] Using a password on the command line interface can be insecure.
Cloning into '/var/www/vhosts/textpattern.co/dev-demo/staging/github'...
remote: Counting objects: 475, done.
remote: Compressing objects: 100% (417/417), done.
remote: Total 475 (delta 99), reused 139 (delta 20), pack-reused 0
Receiving objects: 100% (475/475), 1.36 MiB | 0 bytes/s, done.
Resolving deltas: 100% (99/99), done.
Checking connectivity... done.
[OK]	Connected
[OK]	Using devdemo. (utf8mb4)
[OK]	Creating database tables
[OK]	Lang: 'en'
[OK]	Front-side theme: '/setup/themes/zero'
[OK]	That went well!
mysql: [Warning] Using a password on the command line interface can be insecure.
```

Score. Two for two! And finally, a bogus theme name…setting `"public_theme": "/setup/themes/blokestep",`:

```
$ sudo /var/www/vhosts/textpattern.co/dev-demo/staging/dev-demo.sh
mysql: [Warning] Using a password on the command line interface can be insecure.
Cloning into '/var/www/vhosts/textpattern.co/dev-demo/staging/github'...
remote: Counting objects: 475, done.
remote: Compressing objects: 100% (417/417), done.
remote: Total 475 (delta 99), reused 139 (delta 20), pack-reused 0
Receiving objects: 100% (475/475), 1.36 MiB | 0 bytes/s, done.
Resolving deltas: 100% (99/99), done.
Checking connectivity... done.
[OK]	Connected
[OK]	Using devdemo. (utf8mb4)
[OK]	Creating database tables
[OK]	Lang: 'en'
[OK]	Prefs: 'data/core'
[OK]	Import: 'data/txp_category'
[OK]	Import: 'data/txp_link'
[OK]	Import: 'data/txp_section'
[OK]	Import: 'articles/articles.welcome'
[OK]	Front-side theme: '/setup/themes/four-point-seven'
[OK]	That went well!
mysql: [Warning] Using a password on the command line interface can be insecure.
```
That fails silently (in non-debug mode at least) and installs `four-point-seven` with `[OK]` confirmation.

**Comment by Bloke** [comment]
Thanks for checking it works for now. Not sure what to do about the "no theme exists" situation. We could maybe abort, but I think by the time it gets to that part of the setup process, the remaining tables have already been created and populated. To drop out at that late stage would mean a little bit of unpicking (delete tables by hand, primarily) or - worst case if you didn't notice or it was part of a wider deployment - you'd have an installation in a part-finished state, with no assets and themes.

I'm inclined to think the safest option if there's a minor configuration error is to just breeze on regardless, and install a template we know exists so at least the admin gets a complete admin experience, even if it's the wrong one, They can always then manually pull in the intended theme from the UI, or correct the script, blat the tables/`config.php` and try again.

So the only thing that remains is to figure out what to do with the path. Will devote some thinking time later, but ideas welcome if anyone has any.

**Comment by petecooper** [comment]
I'm tending to agree on all counts, @Bloke – a known-good, contentful (ugh) theme is better than an empty one. @makss might have some ideas on how to tweak the (excellent) auto-installer to handle this situation, seeing as now the GUI approach seems to have pretty much all of the kinks ironed out.

**Comment by petecooper** [comment]
@bloke @philwareham - this looks OK to close. OP issues look resolved, as far as I can tell (and I apologise for turning it into a support thread – appreciate your patience on that).

**Comment by philwareham** [comment]
Wasn’t there a path to figure out? Sorry I’m not too clued up on this issue since I haven’t used the auto-install feature.

**Comment by petecooper** [comment]
@philwareham - resolved in https://github.com/textpattern/textpattern/issues/1101#issuecomment-357919543 I think

**Comment by philwareham** [comment]
Ok thanks - closing.
